src/dbnode/storage/index/block.go

```diff
@@ -501,7 +445,30 @@ func (b *block) QueryIter(ctx context.Context, query Query) (doc.QueryDocIterato
 func (b *block) QueryWithIter(
 	ctx context.Context,
 	opts QueryOptions,
-	docIter doc.Iterator,
+	docIter doc.QueryDocIterator,
+	results DocumentResults,
+	limit int,
+	logFields []opentracinglog.Field,
+) error {
+	ctx, sp := ctx.StartTraceSpan(tracepoint.BlockQuery)
+	sp.LogFields(logFields...)
+	defer sp.Finish()
+
+	err := b.queryWithSpan(ctx, opts, docIter, results, limit)
+	if err != nil {
+		sp.LogFields(opentracinglog.Error(err))
+	}
+	if docIter.Done() {
+		b.metrics.queryDocsMatched.RecordValue(float64(results.TotalDocsCount()))
+		b.metrics.querySeriesMatched.RecordValue(float64(results.Size()))
+	}
```

Comment:
Hm, shouldn't this be done at an end of a query? Not end of querying a block, otherwise the same results will be compoundingly recorded again.

Reply:
good catch. I ended up tracking the counts in the iterators.